src/lexer/jsx.ts

```diff
@@ -64,12 +63,46 @@ export function scanJSXToken(parser: ParserState, context: Context): Token {
   switch (token) {
     // '<'
     case Token.LessThan: {
-      advanceChar(parser);
-      if (parser.currentChar === Chars.Slash) {
-        advanceChar(parser);
-        parser.setToken(Token.JSXClose);
+      nextToken(parser, context);
+      if (parser.getToken() === Token.JSXClose) break;
+      if (parser.getToken() !== Token.LessThan)
+        report(parser, Errors.UnexpectedToken, KeywordDescTable[parser.getToken() & Token.Type]);
+
+      const {
+        index,
+        line,
+        column,
+        tokenPos,
+        startPos,
+        startColumn,
+        startLine,
+        colPos,
+        linePos,
+        currentChar,
+        tokenValue,
+        tokenRaw,
+        tokenRegExp
+      } = parser;
+      const tokenAfter = scanSingleToken(parser, context, LexerState.None);
+
+      if (tokenAfter === Token.Divide) {
+        // Rewrite LessThan + Divide as JSXClose
+        parser.setToken(Token.JSXClose, true);
       } else {
-        parser.setToken(Token.LessThan);
+        // Reset if not merged
+        parser.index = index;
+        parser.line = line;
+        parser.column = column;
+        parser.tokenPos = tokenPos;
+        parser.startPos = startPos;
+        parser.startColumn = startColumn;
+        parser.startLine = startLine;
+        parser.colPos = colPos;
+        parser.linePos = linePos;
+        parser.currentChar = currentChar;
+        parser.tokenValue = tokenValue;
+        parser.tokenRaw = tokenRaw;
+        parser.tokenRegExp = tokenRegExp;
       }
```

Comment:
This is the only way I could come out. Reuse the standard scanner so white spaces and comments will be handled.

Token "<" and "/" need to be merged into "</". Otherwise just return "<" (done by first nextToken() call), and revert effect of the extra scanSingleToken.
In short, we peek ahead and restore if needed.

Reply:
Can we save these to a object (Object.pick())? and restore with `Object.assign()` later? too many to me.

Or add `save()` `restore()`  methods to Parse, like html's ContentRendering2D

Reply:
Maybe we can use `using`? I heard that TS already support.

Reply:
Oh, fancy latest resource management :-)
I will sit with the PR for few days until we are satisfied.

Reply:
I played with jsx yesterday while I ported it to typescript from my assembly code. 

Tricky things to remember is that jsx has it's own special comments, jsx comments and some other nasty features that no one implement. 

You also need to preserve collapsed jsx whitespace.  

The >> is an edge case that is poorly handled by meriyah. Running the normal scanner for jsx will give issue with some punctuators. 

Use a separate scanner

Reply:
Maybe two new APIs to ParserState, like `parser.snapshot(): ParserStateSnapshot` and `parser.restore(snapshot: ParserStateSnapshot)`.
Or `parser.createCheckPoint(): void` and `parser.restoreCheckPoint(): void`?

Reply:
What is snapshot?

Reply:
I got inspiration from meriyah lexer in my rewrite - build the same way  -  but I changed the approach a little so you can check ahead on next token without any backtracking and perf loss. That way you could check if token + 1 is > for >> edge cases. 

Or simply refactor tokens and return a token.jsx node from lexer itself. Similar approach I did for solving ?. edge case

Reply:
You can  change meriyah to return token from sst like this and then handle ast nodes like arrays instead of object and you can do x[1] to get what you want. 

Example

tokens.push({
      type: TokenType.NUMBER,
      value: result,
      loc: ((state.tokenStartLine << 20) | (state.tokenStartColumn << 10) | ((state.lineNumber << 20) | state.column)) >>> 0,
      flags: state.tokenFlags,
    });

Reply:
> Maybe two new APIs to ParserState, like `parser.snapshot(): ParserStateSnapshot` and `parser.restore(snapshot: ParserStateSnapshot)`. Or `parser.createCheckPoint(): void` and `parser.restoreCheckPoint(): void`?

This sounds good

Reply:
To continue on updated code.

Reply:
@3cp if you can wait a week or something then I can show you an example on correctly implementation. 

I'm busy as of now fine in tuning my  type system for my parser - ai insperated type system so no need for TS. 

BTW! JSX has it's own annotations and comments. The comment call back in meriya should be adjusted to handle them. 

And some of this comments obviously gives instruction on how to add additiiycide